internal/plugincommon/consensus/oracle_min_observation.go

```diff
@@ -0,0 +1,68 @@
+package consensus
+
+import (
+	"fmt"
+
+	"github.com/smartcontractkit/libocr/commontypes"
+	"golang.org/x/crypto/sha3"
+
+	cciptypes "github.com/smartcontractkit/chainlink-ccip/pkg/types/ccipocr3"
+)
+
+type observersCounter[T any] struct {
+	data      T
+	observers map[commontypes.OracleID]struct{}
+}
+
+// OracleMinObservation provides a way to ensure a minimum number of observations for
+// some piece of data have occurred. It maintains an internal cache and provides a list
+// of valid or invalid data points.
+type OracleMinObservation[T any] interface {
+	Add(data T, oracleID commontypes.OracleID)
+	GetValid() []T
+}
+
+// minObservation is a helper object to filter data based on observation counts.
+// It keeps track of all inputs, determines if they are consistent
+// with one another, and whether they meet the required count threshold.
+type oracleMinObservation[T any] struct {
+	minObservation Threshold
+	cache          map[cciptypes.Bytes32]*observersCounter[T]
+	idFunc         func(T) [32]byte
+}
+
+// NewOracleMinObservation constructs a concrete MinObservation object. The
+// supplied idFunc is used to generate a uniqueID for the type being observed.
```

Comment:
This doc comment can be improved by removing redundancies

```suggestion
// NewOracleMinObservation returns an OracleMinObservation with the
// supplied idFunc is used to generate a uniqueID for the type being observed.
// If idFunc is nil a default is used.
```